coreclr fails final linking on FreeBSD due to PIC-errors

Attempting to build coreclr gets the build to almost 100%, where it fails with the following error:

````
$ VERBOSE=1 ./build.sh
...
Linking CXX shared library libcoreclr.so
cd /home/josteink/build/cc/bin/obj/FreeBSD.x64.Debug/src/dlls/mscoree/coreclr && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/coreclr.dir/link.txt --verbose=1
/usr/local/bin/clang++35  -fPIC  -Wall -Wno-null-conversion -std=c++11 -nostdinc++ -g -O0  -shared -Wl,-soname,libcoreclr.so -o libcoreclr.so CMakeFiles/coreclr.dir/__/mscoree.cpp.o CMakeFiles/coreclr.dir/__/unixinterface.cpp.o ../../../utilcode/dyncrt/libutilcode.a ../../../debug/ee/wks/libcordbee_wks.a ../../../debug/debug-pal/libdebug-pal.a ../../../unwinder/wks/libunwinder_wks.a ../../../vm/wks/libcee_wks.a ../../../md/compiler/wks/libmdcompiler_wks.a ../../../md/runtime/wks/libmdruntime_wks.a ../../../md/enc/wks/libmdruntimerw_wks.a ../../../md/hotdata/full/libmdhotdata_full.a ../../../classlibnative/bcltype/libbcltype.a ../../../md/ceefilegen/libceefgen.a ../../../jit/libClrJit.a ../../../classlibnative/float/libcomfloat_wks.a ../../../inc/libcorguids.a ../../../gcinfo/libgcinfo.a ../../../debug/ildbsymlib/libildbsymlib.a ../../../strongname/api/wks/libstrongname_wks.a ../../../utilcode/dyncrt/libutilcode.a ../../../binder/libv3binder.a ../../../pal/src/libcoreclrpal.so ../../../palrt/libpalrt.a -lpthread -lrt /usr/local/lib/libunwind.so /usr/local/lib/libintl.so -Wl,-rpath,/home/josteink/build/cc/bin/obj/FreeBSD.x64.Debug/src/pal/src:/usr/local/lib:
clang: warning: argument unused during compilation: '-nostdinc++'
/usr/bin/ld: ../../../debug/ee/wks/libcordbee_wks.a(dbghelpers.S.o): relocation R_X86_64_PC32 against `FuncEvalHijackWorker' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
clang: error: linker command failed with exit code 1 (use -v to see invocation)
*** [src/dlls/mscoree/coreclr/libcoreclr.so] Error code 1
````

As I'm reading this `libcordbee_wks` is not built with PIC enabled, but as far as I'm reading the [CMakeLists.txt for wks](https://github.com/dotnet/coreclr/blob/master/src/debug/ee/wks/CMakeLists.txt), it should be:

````cmake
if (WIN32)
...
else ()
add_compile_options(-fPIC)
add_library(cordbee_wks ${CORDBEE_SOURCES_WKS} ../amd64/dbghelpers.S)
endif (WIN32)
````

Any advice is appreciated.

Thanks @janvorli for the hint. That seems to fix it. We got a new linker problem though
```
Linking CXX executable corerun
cd /home/janhenke/coreclr/bin/obj/FreeBSD.x64.Debug/src/coreclr/hosts/unixcorerun && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/corerun.dir/link.txt --verbose=1
/usr/local/bin/clang++35    -Wall -Wno-null-conversion -std=c++11 -g -O0   CMakeFiles/corerun.dir/corerun.cpp.o  -o corerun  -ldl 
/usr/bin/ld: cannot find -ldl
```
What is libdl? Is it being build by coreclr or do we miss a dependency?

--- a/src/coreclr/hosts/unixcorerun/CMakeLists.txt
+++ b/src/coreclr/hosts/unixcorerun/CMakeLists.txt
@@ -15,9 +15,12 @@ add_executable(corerun
     ${CORERUN_SOURCES}
 )
 
-target_link_libraries(corerun 
-    dl
-)
+# FreeBSD implements dlopen in libc
+if(NOT CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
+    target_link_libraries(corerun 
+        dl
+    )
+endif(NOT CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
 
 add_dependencies(corerun
     coreclr

--- a/src/dlls/mscoree/coreclr/CMakeLists.txt
+++ b/src/dlls/mscoree/coreclr/CMakeLists.txt
@@ -13,7 +13,7 @@ else()
 
 add_definitions(-DNO_CRT_INIT)
 
-if(CMAKE_SYSTEM_NAME STREQUAL Linux)
+if(CMAKE_SYSTEM_NAME STREQUAL Linux OR CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
 # This option is necessary to ensure that the overloaded delete operator defined inside
 # of the utilcode will be used instead of the standard library delete operator.
     set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Xlinker -Bsymbolic-functions")
@@ -26,7 +26,7 @@ if(CMAKE_SYSTEM_NAME STREQUAL Linux)
 # These options are used to force every object to be included even if it's unused.
     set(START_WHOLE_ARCHIVE -Wl,--whole-archive)
     set(END_WHOLE_ARCHIVE -Wl,--no-whole-archive) 
-endif(CMAKE_SYSTEM_NAME STREQUAL Linux)
+endif(CMAKE_SYSTEM_NAME STREQUAL Linux OR CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
 
 if(CMAKE_SYSTEM_NAME STREQUAL Darwin)
 # These options are used to force every object to be included even if it's unused.

